drivers/spi/spi_silabs_eusart.c

```diff
@@ -137,21 +177,339 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	if (spi_silabs_eusart_is_dma_enabled_instance(dev)) {
+		if (!device_is_ready(data->dma_chan_tx.dma_dev)) {
+			return -ENODEV;
+		}
+
+		eusartAdvancedSpiInit.TxFifoWatermark = eusartTxFiFoWatermark1Frame;
+		eusartAdvancedSpiInit.RxFifoWatermark = eusartRxFiFoWatermark1Frame;
+
+		if (data->dma_chan_rx.chan_nb < 0) {
+			data->dma_chan_rx.chan_nb =
+				dma_request_channel(data->dma_chan_rx.dma_dev, NULL);
+		}
+
+		if (data->dma_chan_rx.chan_nb < 0) {
+			LOG_ERR("DMA channel request failed");
+			return -EAGAIN;
+		}
+
+		if (data->dma_chan_tx.chan_nb < 0) {
+			data->dma_chan_tx.chan_nb =
+				dma_request_channel(data->dma_chan_tx.dma_dev, NULL);
+		}
+
+		if (data->dma_chan_tx.chan_nb < 0) {
+			dma_release_channel(data->dma_chan_rx.dma_dev, data->dma_chan_rx.chan_nb);
+			data->dma_chan_rx.chan_nb = 0;
+			LOG_ERR("DMA channel request failed");
+			return -EAGAIN;
+		}
+	}
+#endif
 	/* Enable EUSART clock */
-	err = clock_control_on(eusart_config->clock_dev,
-			       (clock_control_subsys_t)&eusart_config->clock_cfg);
-	if (err < 0) {
-		return err;
+	err = clock_control_on(eusart_cfg->clock_dev,
+			       (clock_control_subsys_t)&eusart_cfg->clock_cfg);
+	if (err < 0 && err != -EALREADY) {
+		goto exit;
 	}
 
 	/* Initialize the EUSART */
-	EUSART_SpiInit(eusart_config->base, &eusartInit);
+	EUSART_SpiInit(eusart_cfg->base, &eusartInit);
 
 	data->ctx.config = config;
 
 	return 0;
+
+exit:
+#ifdef CONFIG_SPI_SILABS_EUSART_DMA
+	dma_release_channel(data->dma_chan_rx.dma_dev, data->dma_chan_rx.chan_nb);
+	dma_release_channel(data->dma_chan_tx.dma_dev, data->dma_chan_tx.chan_nb);
+#endif
+	return err;
+}
+
+static inline void spi_silabs_eusart_pm_policy_get(const struct device *dev)
+{
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	pm_policy_state_lock_get(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+}
+
+static inline void spi_silabs_eusart_pm_policy_put(const struct device *dev)
+{
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	pm_policy_state_lock_put(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+}
+
+static int spi_silabs_eusart_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct spi_silabs_eusart_config *eusart_config = dev->config;
+	int ret;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		ret = clock_control_on(eusart_config->clock_dev,
+				       (clock_control_subsys_t)&eusart_config->clock_cfg);
+
+		if (ret == -EALREADY) {
+			/* Ignore if clock is already on */
+			ret = 0U;
+		} else if (ret < 0) {
+			break;
+		}
+
+		pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_DEFAULT);
+
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_SLEEP);
+
+		EUSART_Enable(eusart_config->base, eusartDisable);
+		ret = clock_control_off(eusart_config->clock_dev,
+					(clock_control_subsys_t)&eusart_config->clock_cfg);
+
+		if (ret == -EALREADY) {
+			/* Ignore if clock is already on */
```

Comment:
`Ignore if clock is already off` (BTW, I suggest to drop this comment rather than confusing the reader).